core/templates/components/forms/custom-forms-directives/thumbnail-uploader.component.ts

```diff
@@ -0,0 +1,261 @@
+        this.imageUploadHelperService.getTrustedResourceUrlForThumbnailFilename(
+          this.filename,
+          this.contextService.getEntityType(),
+          this.contextService.getEntityId()));
+    } else {
+      this.editableThumbnailDataUrl = this.placeholderImageDataUrl;
+    }
+  }
+
+  // 'ngOnChanges' is required here to update the thumbnail image
+  // everytime the thumbnail filename changes (eg. draft is discarded).
+  // The trusted resource url for the thumbnail should not be directly
+  // bound to ngSrc because it can cause an infinite digest error.
+  // This watcher is triggered only if the thumbnail filename of the
+  // model changes. It would change for the following operations:
+  // 1. Initial render of the page containing this directive.
+  // 2. When a thumbnail is uploaded.
+  // 3. When a saved draft is discarded.
+  ngOnChanges(changes: SimpleChanges): void {
+    if (
+      changes.filename &&
+      changes.filename.currentValue !== changes.filename.previousValue) {
+      const newValue = changes.filename.currentValue;
+      const previousValue = changes.filename.previousValue;
+      this.filenameChanges(newValue, previousValue);
+    }
+  }
+
+  filenameChanges(newFilename: string, prevFilename: string): void {
+    if (newFilename) {
+      this.editableThumbnailDataUrl = (
+        this.imageUploadHelperService
+          .getTrustedResourceUrlForThumbnailFilename(
+            newFilename,
+            this.contextService.getEntityType(),
+            this.contextService.getEntityId()));
+      this.uploadedImage = this.editableThumbnailDataUrl;
+    }
+    this.thumbnailIsLoading = true;
+  }
+
+  saveThumbnailBgColor(newBgColor: string): void {
+    if (newBgColor !== this.bgColor) {
+      this.updateBgColor(newBgColor);
+    }
+  }
+
+  saveThumbnailImageData(imageURI: string, callback: {
+     (): void; (): void; }): void {
+    this.resampledFile = null;
+    this.resampledFile = (
+      this.imageUploadHelperService.convertImageDataToImageFile(
+        imageURI));
+
+    if (this.resampledFile === null) {
+      this.alertsService.addWarning('Could not get resampled file.');
+      return;
+    }
+    this.postImageToServer(this.resampledFile, callback);
+  }
+
+  postImageToServer(resampledFile: Blob, callback: () => void): void {
+    let form = new FormData();
+    form.append('image', resampledFile);
+    form.append('payload', JSON.stringify({
+      filename: this.tempImageName,
+      filename_prefix: 'thumbnail'
+    }));
+    var imageUploadUrlTemplate = '/createhandler/imageupload/' +
+      '<entity_type>/<entity_id>';
+    this.csrfTokenService.getTokenAsync().then((token) => {
+      form.append('csrf_token', token);
+      $.ajax({
+        url: this.urlInterpolationService.interpolateUrl(
+          imageUploadUrlTemplate, {
+            entity_type: this.contextService.getEntityType(),
+            entity_id: this.contextService.getEntityId()
+          }
```

Comment:
Is there any specific reason to use jquery here? If this can't be done using `HttpClient`, please mention the reason.

Reply:
Will have to initialize Csrf token here again. I tried creating a function in AssetsBackendApiService same as that of `saveMathExpressionImage`but it wasn't working that way.

Reply:
getTokenAsync() doesnot initialize a new token. Csrf token is only initialized once when page is first loaded. getTokenAsync() just provides the same token again and again.
I understand that http interceptor will add csrf token in the request body but here we need to add it to the form data.

@srijanreddy98 what do you think about this?

Reply:
Got to this PR: https://github.com/oppia/oppia/pull/12764/files, find the file svg-filename-editor.component.ts (Note the file may have not been rendered. Search for a function named postSvgTo server.

## Old function.
![Screenshot 2021-05-20 at 7 59 24 AM](https://user-images.githubusercontent.com/23002114/118909903-4feee080-b941-11eb-9d35-3b95101acbd0.png)

## New function:
![Screenshot 2021-05-20 at 8 00 11 AM](https://user-images.githubusercontent.com/23002114/118909962-6a28be80-b941-11eb-8078-1dad5561327e.png)

## Migrated ajax function:
![Screenshot 2021-05-20 at 8 01 12 AM](https://user-images.githubusercontent.com/23002114/118910062-8f1d3180-b941-11eb-8f95-ca4cde6364a6.png)

Reply:
fixed